### PR TITLE
AP_RangeFinder: add support for RDS02UF radar driver on seria

### PR DESCRIPTION
**Pull request by zebulon-86**
## Summary

Resubmit pull request，relevant : https://github.com/ArduPilot/ardupilot/pull/22385
Adding new features
AP_RangeFinder:add support for RDS02UF radar driver on serial。

## Documentation

Product URL：[Microbrain Intelligent](https://www.microbrain.com.cn/en/ProductCenter/list.aspx?categoryid=132)
Related Manuals:[【Benewake(北醒)】毫米波雷达在无人机上避障资料汇总](https://blog.csdn.net/zoran_wu/article/details/126750128)
Here are the dumped Manuals:[RDS02U_Uart_Manual](https://github.com/zebulon-86/Document/blob/master/MicrobrainRadar/RDS02U_Uart_Manual.md)
For PDF:[RDS02U](https://github.com/zebulon-86/Document/tree/master/MicrobrainRadar/pdf)

## Change

Some tweaks to the code related to #22385

[libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h](https://github.com/ArduPilot/ardupilot/pull/22385/files#diff-bf4ef8c6a1897e12756a77e6efd92c249789ccd5788f54610ad9a6d1e6140622)
```c
#if RDS02_USE_CRC
    const uint8_t crc8_table[256] = {
    0x93,0x98,0xE4,0x46,0xEB,0xBA,0x04,0x4C,
```
The CRC table takes up too much flash space,so added the define of whether to enable CRC, which is not enabled by default, and only use the frame header and frame tail to judge the integrity of the frame. It is not a big problem not to do CRC when the serial communication is stable. Users can also choose to enable CRC according to the hardware situation.

### CODE REVIEW

**Comment by magicrub** [comment]
Why not use existing crc functions?

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                            _distance_m = RDS02UF_DIST_MIN_M;
+                            set_status(RangeFinder::Status::OutOfRangeLow);
+                        }
+                        ret = true;
+                        state.last_reading_ms = AP_HAL::millis();
+                    }
+
+                }
+            }
+			parser_index = 0;
```
Comment (muramura) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:204]: ardupilot indentation is whitespace 4.

**Review by muramura — CHANGES_REQUESTED** [review]

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp (hunk 1 above)
Comment (zebulon-86) [reply libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp]: I didn't notice that tabs are used here, I will make a modification, thank you for pointing out the problem

**Comment by rmackay9** [comment]
Hi @zebulon-86,

It looks like this is failing CI because of a problem with a definition.

[ 469/1015] Compiling libraries/AP_RangeFinder/AP_RangeFinder_PulsedLightLRF.cpp
../../libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:155:13: error: "RDS02_USE_CRC" is not defined, evaluates to 0 [-Werror=undef]
  155 |         #if RDS02_USE_CRC
      |             ^~~~~~~~~~~~~

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
this has no metadata update to the rf type param

**Comment by zebulon-86** [comment]
> Hi @zebulon-86,
> 
> It looks like this is failing CI because of a problem with a definition.
> 
> [ 469/1015] Compiling libraries/AP_RangeFinder/AP_RangeFinder_PulsedLightLRF.cpp ../../libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:155:13: error: "RDS02_USE_CRC" is not defined, evaluates to 0 [-Werror=undef] 155 | #if RDS02_USE_CRC | ^~~~~~~~~~~~~

I added in build_options.py
```
Feature('Rangefinder', 'RANGEFINDER_RDS02UF', 'AP_RANGEFINDER_RDS02UF_ENABLED', "Enable Rangefinder - RDS02UF", 0, "RANGEFINDER"), # NOQA: E501
```
Added `#if AP_RANGEFINDER_RDS02UF_ENABLED`  in AP_RangeFinder_RDS02UF.cpp, now All checks have passed
@rmackay9

**Comment by zebulon-86** [comment]
> this has no metadata update to the rf type param

@Hwurzburg 
Sorry, I don't quite understand which one this refers to specifically. Is it the following definition?
```
MAV_DISTANCE_SENSOR _get_mav_distance_sensor_type() const override {
    return MAV_DISTANCE_SENSOR_RADAR;
}
```

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+            decode_state = RDS02UF_PARSE_STATE::STATE6_LENGTH_L;
+            break;
+        case RDS02UF_PARSE_STATE::STATE6_LENGTH_L: // lengh_low
+            parser_buffer[parser_index++] = data;
+            decode_state = RDS02UF_PARSE_STATE::STATE7_LENGTH_H;
+            break;
+        case RDS02UF_PARSE_STATE::STATE7_LENGTH_H: // lengh_high
+        {
+            uint8_t read_len = data << 8 | parser_buffer[parser_index-1];
+            if (read_len == RDS02UF_DATA_LEN)	// rds02uf data length is always 10
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:139]: nitpick: bracket should be on the earlier line

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    float _distance_m;
+
+    // get a distance reading
+    bool get_reading(float &reading_m) override;
+    uint16_t read_timeout_ms() const override { return 3000; }
+    uint8_t crc8(uint8_t *pbuf, int32_t len);
+    bool decode(uint8_t c);
+
+#if RDS02_USE_CRC
+    const uint8_t crc8_table[256] = {
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h:95]: This should be moved to AP_Math's crc file

File: libraries/SITL/SIM_RF_RDS02UF.cpp [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+*/
+
+#include "SIM_RF_RDS02UF.h"
+
+#include <stdio.h>
+
+using namespace SITL;
+
+
+    const uint8_t crc8_table[256] = {
```
Comment (rmackay9) [inline libraries/SITL/SIM_RF_RDS02UF.cpp:26]: we should remove this and use the version in AP_Math once it has been moved.

File: Tools/autotest/arducopter.py [hunk 5]
```diff
@@ -7932,6 +7932,7 @@ def RangeFinderDrivers(self):
             ("benewake_tf03", 27),
             ("gyus42v2", 31),
             ("teraranger_serial", 35),
+            ("rds02uf", 37),
```
Comment (rmackay9) [inline Tools/autotest/arducopter.py:7935]: This should be moved to a separate commit starting with "Tools:".

**Comment by Hwurzburg** [comment]
add metadata in https://github.com/ArduPilot/ardupilot/blob/8653bf7349fcd98e7a38db4af9a73bda83e8e186/libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp#L14

File: libraries/AP_HAL/SIMState.cpp [hunk 6]
```diff
@@ -163,6 +163,9 @@ void SIMState::fdm_input_local(void)
     if (leddarone != nullptr) {
         leddarone->update(sitl_model->rangefinder_range());
     }
+    if (rds02uf != nullptr) {
```
Comment (rmackay9) [inline libraries/AP_HAL/SIMState.cpp:166]: These AP_HAL_SIMState changes should be moved to a separate commit prefixed with "AP_HAL:"

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+
+// add a single character to the buffer and attempt to decode
+// returns true if a distance was successfully decoded
+bool AP_RangeFinder_RDS02UF::decode(uint8_t c)
+{
+    uint8_t data = c;
+    bool ret = false;
+    switch (decode_state)
+    {
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:100]: nitpick: bracket should be one line higher

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+
+    return ret;
+}
+
+uint8_t AP_RangeFinder_RDS02UF::crc8(uint8_t* pbuf, int32_t len)
+{
+#if RDS02_USE_CRC
+     uint8_t crc = 0;
+     uint8_t* data = pbuf;
+     while ( len-- )
```
Comment (rmackay9) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:218]: let's remove the white space between the brackets.

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h (hunk 3 above)
Comment (zebulon-86) [reply libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h]: I submitted a commit and moved the crc table to AP_Math/crc.h

```
#define RDS02_USE_CRC 0
#if RDS02_USE_CRC
    const uint8_t rds02_crc8_table[256] = {
```

Because this table may not be applicable elsewhere, RDS02_USE_CRC is added, and this table is not enabled by default

File: libraries/SITL/SIM_RF_RDS02UF.cpp (hunk 4 above)
Comment (zebulon-86) [reply libraries/SITL/SIM_RF_RDS02UF.cpp]: Already moved

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp (hunk 8 above)
Comment (zebulon-86) [reply libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp]: Thanks for the reminder, I'm not careful enough about the format, I checked again

**Comment by zebulon-86** [comment]
> also need to add to extract_features.py when adding a feature to build server

I added the following code to the extract_features.py file.

```
('AP_RANGEFINDER_RDS02UF_ENABLED', r'AP_RangeFinder_RDS02UF::get_reading\b',),
```

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h (hunk 3 above)
Comment (zebulon-86) [reply libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h:95]: I don't know if it is appropriate to add this crc8 table to AP_Math, because this table is a fixed value defined by the manufacturer and not applicable to other devices. Because there are frame header and frame tail restrictions, there will be no problem in short-distance serial communication if the crc check is not applicable, and this check table can also be deleted directly.

**Comment by peterbarker** [comment]
All of my review comments on https://github.com/ArduPilot/ardupilot/pull/21646 almost certainly still apply.

Please do not open multiple PRs, but force-push over the branch of the previous PR.

**Comment by rishabsingh3003** [comment]
@zebulon-86 I have rebased this PR. I have also fixed the issues mentioned above and made a few changes to the PR. The changes I have made are in a different commit.

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+bool AP_RangeFinder_RDS02UF::get_reading(float &reading_m)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    // read any available data from the radar
+    float sum = 0.0f;
+    uint16_t count = 0;
+    uint32_t nbytes = MAX(uart->available(), 1024U);
```
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:42]: should be MIN() not MAX()

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    float distance_m;
+    while (nbytes-- > 0) {
+        const int16_t c = uart->read();
```
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:45]: should use the bool version of read()

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+            decode_state = RDS02UF_PARSE_STATE::STATE6_LENGTH_L;
+            break;
+        case RDS02UF_PARSE_STATE::STATE6_LENGTH_L:
+            // length_low
+            parser_buffer[parser_index++] = data;
+            decode_state = RDS02UF_PARSE_STATE::STATE7_LENGTH_H;
+            break;
+        case RDS02UF_PARSE_STATE::STATE7_LENGTH_H:{
+            // length_high
+            uint8_t read_len = data << 8 | parser_buffer[parser_index-1];
```
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:113]: how do you get data<<8 into a uint8_t ?

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+            // rds02uf data length is always 10
+            if (read_len == RDS02UF_DATA_LEN) {
```
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:115]: need a check for read_len longer than buffer length of 50

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+                        distance_m = (parser_buffer[RDS02_DATA_Y_INDEX_H] * 256 + parser_buffer[RDS02_DATA_Y_INDEX_L]) * 0.01f;
+                        ret = true;
+                        state.last_reading_ms = AP_HAL::millis();
+                    }
+                }
+            }
+			parser_index = 0;
+			decode_state = RDS02UF_PARSE_STATE::STATE0_SYNC_1;
+			break;
+        }
```
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:170]: check index == sizeof(buffer), and reset at this point, so we don't put any more bytes in

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+#pragma once
+
+#include "AP_RangeFinder_config.h"
+
+#if AP_RANGEFINDER_RDS02UF_ENABLED
+
+#include "AP_RangeFinder.h"
+#include "AP_RangeFinder_Backend_Serial.h"
+
+#define RDS02_BUFFER_SIZE           50
```
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h:25]: move most defines to cpp file

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by rishabsingh3003** [comment]
From Dev Call discussion - Redo the implementation with structure

**Comment by tridge** [comment]
@peterbarker will review parser and merge

**Comment by rishabsingh3003** [comment]
Added "memchr"

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define RDS02_HEAD1                 0x55
+#define RDS02_HEAD2                 0x55
+#define RDS02_END                   0xAA
+#define RDS02UF_PRE_DATA_LEN        6
+#define RDS02_TARGET_FC_INDEX_L     8
+#define RDS02_TARGET_FC_INDEX_H     9
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:37]: These aren't used now

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    return true;
+}
+
+// find a RDS02UF message in the buffer, starting at
+// initial_offset.  If found, that message (or partial message) will
+// be moved to the start of the buffer.
+void AP_RangeFinder_RDS02UF::move_header_in_buffer(uint8_t initial_offset)
+{
+    uint8_t* header_ptr = (uint8_t*)memchr(&u.parse_buffer[initial_offset], RDS02_HEAD1, body_length - initial_offset);
+    if (header_ptr != nullptr) {
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:116]: This doesn't look right.  In the case we don't find anything we won't empty the buffer?

AP_ExternalAHRS_VectorNAV does this, for example.

**Review by peterbarker — CHANGES_REQUESTED** [review]
Please test this with a little bit of corruption, make sure it still generally works (AP_SIM_SERIALDEVICE_CORRUPTION_ENABLED).  Run under valgrind when doing this.

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp (hunk 16 above)
Comment (rishabsingh3003) [reply libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:111]: Thanks for finding this!

**Comment by rishabsingh3003** [comment]
rebased, and fixed the pending comments

**Comment by rishabsingh3003** [comment]
Also tested with, AP_SIM_SERIALDEVICE_CORRUPTION_ENABLED and valgrind. Works well.
Tested with real sensor connected to SITL as well.

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+        const uint16_t read_info_fc = (u.packet.data[1] << 8 | u.packet.data[0]);
+        if ((read_info_fc & RDS02UF_IGNORE_ID_BYTE) == RDS02_TARGET_INFO_FC) {
+            // read_info_fc = 0x70C + ID * 0x10, ID: 0~0xF
+            distance_m = (u.packet.data[6] * 256 + u.packet.data[5]) * 0.01f;
+            state.last_reading_ms = AP_HAL::millis();
+        }
+    }
+
+    // reset buffer
+    body_length = 0;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:101]:
```suggestion
    // consume packet
    move_header_in_buffer(sizeof(u.packet));
```

The current fixed size makes this NFC, but if we wanted to be able to read multiple packets at once this would be required....

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+                                 ARRAY_SIZE(u.parse_buffer)-body_length);
+    if (nbytes == 0) {
+        return false;
+    }
+    body_length += nbytes;
+
+    move_header_in_buffer(0);
+
+    // header byte 1 is correct.
+    if (body_length < ARRAY_SIZE(u.parse_buffer)) {
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.cpp:55]:
```suggestion
    if (body_length < sizeof(u.packet)) {
```

... this is, after all, exactly what you're interested in.

The parse buffer can be larger than the packet, allowing for multiple packets to be read in at once (and different packets to be parsed out of the same buffer, of course)

File: libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+        uint8_t length_h;
+        uint8_t data[RDS02UF_DATA_LEN];
+        uint8_t checksum;
+        uint8_t footermagic1;
+        uint8_t footermagic2;
+    };
+
+    union RDS02UF_Union {
+        uint8_t parse_buffer[21];
+        struct RDS02UFPacket packet;
```
Comment (peterbarker) [inline libraries/AP_RangeFinder/AP_RangeFinder_RDS02UF.h:95]: Technically we really should mark this as `PACKED`.  Since all members are 8-byte quantities it doesn't really matter.

**Review by peterbarker — APPROVED** [review]

**Comment by peterbarker** [comment]
Merged, thanks all!

**Comment by Hwurzburg** [comment]
@zebulon-86 I can find no vendor offering this device and Benewake does not list on its web page.....I cant see any reason to have a wiki entry for this device since its not available

**Comment by rishabsingh3003** [comment]
@Hwurzburg I believe Benewake plans to sell these. We have a chat with them where we could ask

**Comment by Hwurzburg** [comment]
@zebulon-86 as far as I can tell this product does not exist and cannot be obtained......no wiki page will be generated for it at this time due to this

**Comment by Hwurzburg** [comment]
@rishabsingh3003 I see no chat in Skype or Discord for them.....if someone can give me a link as to where to obtain this device, I will create a wiki page and entry then
